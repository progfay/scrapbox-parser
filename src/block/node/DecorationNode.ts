import { NodeParserType, LineNodeType, convertToLineNodes } from '.'

const decorationRegExp = /^(.*?)\[([!"#%&'()*+,-./{|}<>_~]+) ((?:\[[^\]]+\]|[^\]])+)\](.*)$/

export type DecorationCharType = '*' | '!' | '"' | '#' | '%' | '&' | '\'' | '(' | ')' | '+' | ',' | '-' | '.' | '/' | '{' | '|' | '}' | '<' | '>' | '_' | '~'
export type AsteriskDecorationCharType = '*-1' | '*-2' | '*-3' | '*-4' | '*-5' | '*-6' | '*-7' | '*-8' | '*-9' | '*-10'
export type DecorationType = Exclude<DecorationCharType, '*'> | AsteriskDecorationCharType
export type DecorationNodeType = {
  readonly type: 'decoration'
  readonly decos: ReadonlyArray<DecorationType>
  readonly nodes: ReadonlyArray<LineNodeType>
}

const createDecorationNode = (decoChars: string, nodes: LineNodeType[]): DecorationNodeType => {
  const decoSet = new Set<string>(decoChars)
  if (decoSet.has('*')) {
    const asteriskCount = decoChars.split('*').length - 1
    decoSet.delete('*')
    decoSet.add(`*-${Math.min(asteriskCount, 10)}` as DecorationCharType)
  }

  return {
    type: 'decoration',
    decos: Array.from(decoSet) as DecorationType[],
    nodes
  }
}

export const DecorationNodeParser: NodeParserType = (text, { nested, quoted }, next) => {
  if (nested) return next()

  const decorationMatch = text.match(decorationRegExp)
  if (!decorationMatch) return next()

  const [, left, decoChars, target, right] = decorationMatch
  return [
    ...convertToLineNodes(left, { nested, quoted }),
    createDecorationNode(decoChars, convertToLineNodes(target, { nested: true, quoted })),
    ...convertToLineNodes(right, { nested, quoted })
  ]
}
