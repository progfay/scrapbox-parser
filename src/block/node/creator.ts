import { convertToNodes } from "./index.ts";

import type { NodeParser, NodeParserOption } from "./index.ts";
import type { Node } from "./type.ts";

export type NodeCreator<T extends Node> = (
	target: string,
	opts: NodeParserOption,
) => T[];

type NodeParserCreator<T extends Node> = (
	nodeCreator: NodeCreator<T>,
	opts: { parseOnNested: boolean; parseOnQuoted: boolean; patterns: RegExp[] },
) => NodeParser;

export const createNodeParser: NodeParserCreator<Node> = (
	nodeCreator,
	{ parseOnNested, parseOnQuoted, patterns },
) => {
	return (text, opts, next) => {
		if (!parseOnNested && opts.nested) return next?.() ?? [];
		if (!parseOnQuoted && opts.quoted) return next?.() ?? [];

		for (const pattern of patterns) {
			const match = pattern.exec(text);
			if (match === null) continue;

			const left = text.substring(0, match.index);
			const right = text.substring(match.index + (match[0]?.length ?? 0));

			const node = nodeCreator(match[0] ?? "", opts);
			return [
				...convertToNodes(left, opts),
				...node,
				...convertToNodes(right, opts),
			];
		}

		return next?.() ?? [];
	};
};
