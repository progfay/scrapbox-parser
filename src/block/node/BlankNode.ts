import { createNodeParser } from "./creator";
import { createPlainNode } from "./PlainNode";

import type { NodeParser } from "./index";
import type { BlankNode, PlainNode } from "./type";
import type { NodeCreator } from "./creator";

const blankRegExp = /\[\s+\]/;

const createBlankNode: NodeCreator<BlankNode | PlainNode> = (raw, opts) =>
  opts.context === "table"
    ? createPlainNode(raw, opts)
    : [
        {
          type: "blank",
          raw,
          text: raw.substring(1, raw.length - 1),
        },
      ];

export const BlankNodeParser: NodeParser = createNodeParser(createBlankNode, {
  parseOnNested: false,
  parseOnQuoted: true,
  patterns: [blankRegExp],
});
