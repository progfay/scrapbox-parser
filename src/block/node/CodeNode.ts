import { createPlainNode } from "./PlainNode";
import { createNodeParser } from "./creator";

import type { NodeCreator } from "./creator";
import type { NodeParser } from "./index";
import type { CodeNode, PlainNode } from "./type";

const codeRegExp = /`.*?`/;

const createCodeNode: NodeCreator<CodeNode | PlainNode> = (raw, opts) =>
	opts.context === "table"
		? createPlainNode(raw, opts)
		: [
				{
					type: "code",
					raw,
					text: raw.substring(1, raw.length - 1),
				},
			];

export const CodeNodeParser: NodeParser = createNodeParser(createCodeNode, {
	parseOnNested: false,
	parseOnQuoted: true,
	patterns: [codeRegExp],
});
