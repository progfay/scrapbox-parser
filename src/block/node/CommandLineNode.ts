import { createPlainNode } from "./PlainNode.ts";
import { createNodeParser } from "./creator.ts";

import type { NodeCreator } from "./creator.ts";
import type { NodeParser } from "./index.ts";
import type { CommandLineNode, PlainNode } from "./type.ts";

const commandLineRegExp = /^[$%] .+$/;

const createCommandLineNode: NodeCreator<CommandLineNode | PlainNode> = (
	raw: string,
	opts,
) => {
	if (opts.context === "table") {
		return createPlainNode(raw, opts);
	}

	const symbol = raw[0] ?? "";
	const text = raw.substring(2);

	return [
		{
			type: "commandLine",
			raw,
			symbol,
			text,
		},
	];
};

export const CommandLineNodeParser: NodeParser = createNodeParser(
	createCommandLineNode,
	{
		parseOnNested: false,
		parseOnQuoted: false,
		patterns: [commandLineRegExp],
	},
);
