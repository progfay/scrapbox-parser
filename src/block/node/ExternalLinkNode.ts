import { createNodeParser } from "./creator";
import { createPlainNode } from "./PlainNode";

import type { NodeParser } from "./index";
import type { NodeCreator } from "./creator";
import type { LinkNode, PlainNode } from "./type";

const hrefFirstUrlRegExp = /\[https?:\/\/[^\s\]]+\s+[^\]]*[^\s]\]/;
const contentFirstUrlRegExp = /\[[^[\]]*[^\s]\s+https?:\/\/[^\s\]]+\]/;
const bracketedUrlRegExp = /\[https?:\/\/[^\s\]]+\]/;
const httpRegExp = /https?:\/\/[^\s]+/;

const createExternalLinkNode: NodeCreator<LinkNode | PlainNode> = (
  raw,
  opts,
) => {
  if (opts.context === "table") {
    return createPlainNode(raw, opts);
  }

  const inner =
    raw.startsWith("[") && raw.endsWith("]")
      ? raw.substring(1, raw.length - 1)
      : raw;

  const isHrefFirst = /^https?:\/\/[^\s\]]/.test(inner);
  const match = (
    isHrefFirst ? /^https?:\/\/[^\s\]]+/ : /https?:\/\/[^\s\]]+$/
  ).exec(inner);
  if (match?.[0] === undefined) return [];

  const content = isHrefFirst
    ? inner.substring(match[0].length)
    : inner.substring(0, match.index - 1);

  return [
    {
      type: "link",
      raw,
      pathType: "absolute",
      href: match[0],
      content: content.trim(),
    },
  ];
};

export const ExternalLinkNodeParser: NodeParser = createNodeParser(
  createExternalLinkNode,
  {
    parseOnNested: true,
    parseOnQuoted: true,
    patterns: [
      hrefFirstUrlRegExp,
      contentFirstUrlRegExp,
      bracketedUrlRegExp,
      httpRegExp,
    ],
  },
);
