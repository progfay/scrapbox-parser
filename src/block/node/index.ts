import { QuoteNodeParser } from './QuoteNode'
import { HelpfeelNodeParser } from './HelpfeelNode'
import { StrongImageNodeParser } from './StrongImageNode'
import { StrongIconNodeParser } from './StrongIconNode'
import { StrongNodeParser } from './StrongNode'
import { FormulaNodeParser } from './FormulaNode'
import { DecorationNodeParser } from './DecorationNode'
import { CodeNodeParser } from './CodeNode'
import { CommandLineNodeParser } from './CommandLineNode'
import { BlankNodeParser } from './BlankNode'
import { ImageNodeParser } from './ImageNode'
import { ExternalLinkNodeParser } from './ExternalLinkNode'
import { GoogleMapNodeParser } from './GoogleMapNode'
import { InternalLinkNodeParser } from './InternalLinkNode'
import { IconNodeParser } from './IconNode'
import { HashTagNodeParser } from './HashTagNode'
import { PlainNodeParser } from './PlainNode'

import type { Node } from './type'

export interface NodeParserOption {
  nested: boolean
  quoted: boolean
}
export type NextNodeParser = () => Node[]
export type NodeParser = (text: string, opts: NodeParserOption, next?: NextNodeParser) => Node[]

const FalsyEliminator: NodeParser = (text, _, next) => {
  if (text === '') return []
  return next?.() ?? []
}

const combineNodeParsers = (...parsers: NodeParser[]) => (
  text: string = '',
  opts: NodeParserOption = { nested: false, quoted: false }
): Node[] =>
  parsers.reduceRight(
    (acc: NextNodeParser, parser: NodeParser): NextNodeParser => () => parser(text, opts, acc),
    () => PlainNodeParser(text, opts)
  )()

export const convertToNodes = combineNodeParsers(
  FalsyEliminator,
  QuoteNodeParser,
  HelpfeelNodeParser,
  CodeNodeParser,
  CommandLineNodeParser,
  FormulaNodeParser,
  BlankNodeParser,
  DecorationNodeParser,
  StrongImageNodeParser,
  StrongIconNodeParser,
  StrongNodeParser,
  ImageNodeParser,
  ExternalLinkNodeParser,
  IconNodeParser,
  GoogleMapNodeParser,
  InternalLinkNodeParser,
  HashTagNodeParser
)
