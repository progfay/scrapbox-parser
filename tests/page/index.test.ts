import * as fs from "node:fs";
import path from "node:path";
import { describe, it } from "node:test";
import { parse } from "../../src/index.ts";

describe("page", () => {
	it("Empty page", ({ assert }) => {
		const input = "";
		assert.snapshot(parse(input, { hasTitle: true }));
	});

	it("Title Block without `hasTitle` option", ({ assert }) => {
		const input = "Title";
		assert.snapshot(parse(input));
	});

	it("https://scrapbox.io/help/Syntax", ({ assert }) => {
		const input = fs
			.readFileSync(path.resolve("tests/page/input.txt"))
			.toString();
		assert.snapshot(parse(input, { hasTitle: true }));
	});
});
