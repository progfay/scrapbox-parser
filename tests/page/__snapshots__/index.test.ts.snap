// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page Empty page 1`] = `
[
  {
    "text": "",
    "type": "title",
  },
]
`;

exports[`page Title Block without \`hasTitle\` option 1`] = `
[
  {
    "text": "Title",
    "type": "title",
  },
]
`;

exports[`page https://scrapbox.io/help/Syntax 1`] = `
[
  {
    "text": "Syntax",
    "type": "title",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "link": "",
        "raw": "[https://gyazo.com/0f82099330f378fe4917a1b4a5fe8815]",
        "src": "https://gyazo.com/0f82099330f378fe4917a1b4a5fe8815/thumb/1000",
        "type": "image",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Internal Links",
            "text": "Internal Links",
            "type": "plain",
          },
        ],
        "raw": "[[Internal Links]]",
        "type": "strong",
      },
      {
        "raw": " (linking to another page on scrapbox)",
        "text": " (linking to another page on scrapbox)",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[link]\`",
        "text": "[link]",
        "type": "code",
      },
      {
        "raw": " ⇒ ",
        "text": " ⇒ ",
        "type": "plain",
      },
      {
        "content": "",
        "href": "Link",
        "pathType": "relative",
        "raw": "[Link]",
        "type": "link",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "External  Links",
            "text": "External  Links",
            "type": "plain",
          },
        ],
        "raw": "[[External  Links]]",
        "type": "strong",
      },
      {
        "raw": " (linking to another web page)",
        "text": " (linking to another web page)",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`http://google.com\`",
        "text": "http://google.com",
        "type": "code",
      },
      {
        "raw": " ⇒ ",
        "text": " ⇒ ",
        "type": "plain",
      },
      {
        "content": "",
        "href": "http://google.com",
        "pathType": "absolute",
        "raw": "http://google.com",
        "type": "link",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[http://google.com Google]\`",
        "text": "[http://google.com Google]",
        "type": "code",
      },
      {
        "raw": " ⇒ ",
        "text": " ⇒ ",
        "type": "plain",
      },
      {
        "content": "Google",
        "href": "http://google.com",
        "pathType": "absolute",
        "raw": "[http://google.com Google]",
        "type": "link",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "raw": "or",
        "text": "or",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[Google http://google.com]\`",
        "text": "[Google http://google.com]",
        "type": "code",
      },
      {
        "raw": " ⇒ ",
        "text": " ⇒ ",
        "type": "plain",
      },
      {
        "content": "Google",
        "href": "http://google.com",
        "pathType": "absolute",
        "raw": "[Google http://google.com]",
        "type": "link",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Images",
            "text": "Images",
            "type": "plain",
          },
        ],
        "raw": "[[Images]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Direct image link ↓",
        "text": "Direct image link ↓",
        "type": "plain",
      },
      {
        "raw": "\`[https://gyazo.com/da78df293f9e83a74b5402411e2f2e01.png]\`",
        "text": "[https://gyazo.com/da78df293f9e83a74b5402411e2f2e01.png]",
        "type": "code",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "link": "",
        "raw": "[https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png]",
        "src": "https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png",
        "type": "image",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Clickable Thumbnail Links",
            "text": "Clickable Thumbnail Links",
            "type": "plain",
          },
        ],
        "raw": "[[Clickable Thumbnail Links]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "↓ ",
        "text": "↓ ",
        "type": "plain",
      },
      {
        "raw": "\`[http://cutedog.com https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png]\`",
        "text": "[http://cutedog.com https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png]",
        "type": "code",
      },
      {
        "raw": " ",
        "text": " ",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "link": "http://cutedog.com",
        "raw": "[http://cutedog.com https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png]",
        "src": "https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png",
        "type": "image",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Adding the link at the end also works, as before:",
        "text": "Adding the link at the end also works, as before:",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 2,
    "nodes": [
      {
        "raw": "\`[https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png http://cutedog.com]\`",
        "text": "[https://i.gyazo.com/da78df293f9e83a74b5402411e2f2e01.png http://cutedog.com]",
        "type": "code",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Linking to other scrapbox projects",
            "text": "Linking to other scrapbox projects",
            "type": "plain",
          },
        ],
        "raw": "[[Linking to other scrapbox projects]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[/projectname/pagename]\`",
        "text": "[/projectname/pagename]",
        "type": "code",
      },
      {
        "raw": " ⇛ ",
        "text": " ⇛ ",
        "type": "plain",
      },
      {
        "content": "",
        "href": "/icons/check",
        "pathType": "root",
        "raw": "[/icons/check]",
        "type": "link",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[/projectname]\`",
        "text": "[/projectname]",
        "type": "code",
      },
      {
        "raw": " ⇛ ",
        "text": " ⇛ ",
        "type": "plain",
      },
      {
        "content": "",
        "href": "/icons",
        "pathType": "root",
        "raw": "[/icons]",
        "type": "link",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Icons",
            "text": "Icons",
            "type": "plain",
          },
        ],
        "raw": "[[Icons]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[ben.icon]\`",
        "text": "[ben.icon]",
        "type": "code",
      },
      {
        "raw": " ⇛  ",
        "text": " ⇛  ",
        "type": "plain",
      },
      {
        "path": "ben",
        "pathType": "relative",
        "raw": "[ben.icon]",
        "type": "icon",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[/icons/todo.icon]\`",
        "text": "[/icons/todo.icon]",
        "type": "code",
      },
      {
        "raw": " ⇛ ",
        "text": " ⇛ ",
        "type": "plain",
      },
      {
        "path": "/icons/todo",
        "pathType": "root",
        "raw": "[/icons/todo.icon]",
        "type": "icon",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Bold text",
            "text": "Bold text",
            "type": "plain",
          },
        ],
        "raw": "[[Bold text]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[[Bold]]\`",
        "text": "[[Bold]]",
        "type": "code",
      },
      {
        "raw": " or ",
        "text": " or ",
        "type": "plain",
      },
      {
        "raw": "\`[* Bold]\`",
        "text": "[* Bold]",
        "type": "code",
      },
      {
        "raw": "⇒ ",
        "text": "⇒ ",
        "type": "plain",
      },
      {
        "nodes": [
          {
            "raw": "Bold",
            "text": "Bold",
            "type": "plain",
          },
        ],
        "raw": "[[Bold]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Italic text",
            "text": "Italic text",
            "type": "plain",
          },
        ],
        "raw": "[[Italic text]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[/ italic]\`",
        "text": "[/ italic]",
        "type": "code",
      },
      {
        "raw": "⇛ ",
        "text": "⇛ ",
        "type": "plain",
      },
      {
        "decos": [
          "/",
        ],
        "nodes": [
          {
            "raw": "italic",
            "text": "italic",
            "type": "plain",
          },
        ],
        "raw": "[/ italic]",
        "rawDecos": "/",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": " Strikethrough text",
            "text": " Strikethrough text",
            "type": "plain",
          },
        ],
        "raw": "[[ Strikethrough text]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[- strikethrough]\`",
        "text": "[- strikethrough]",
        "type": "code",
      },
      {
        "raw": "⇛ ",
        "text": "⇛ ",
        "type": "plain",
      },
      {
        "decos": [
          "-",
        ],
        "nodes": [
          {
            "raw": "strikethrough",
            "text": "strikethrough",
            "type": "plain",
          },
        ],
        "raw": "[- strikethrough]",
        "rawDecos": "-",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "link": "",
        "raw": "[https://gyazo.com/00ab07461d502db91c8ae170276d1396]",
        "src": "https://gyazo.com/00ab07461d502db91c8ae170276d1396/thumb/1000",
        "type": "image",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Bullet points",
            "text": "Bullet points",
            "type": "plain",
          },
        ],
        "raw": "[[Bullet points]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Press space or tab on a new line to indent and create a bullet point",
        "text": "Press space or tab on a new line to indent and create a bullet point",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 2,
    "nodes": [
      {
        "raw": "Press backspace to remove the indent  / bullet point",
        "text": "Press backspace to remove the indent  / bullet point",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Internal links serve triple duty",
            "text": "Internal links serve triple duty",
            "type": "plain",
          },
        ],
        "raw": "[[Internal links serve triple duty]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "\`[Links]\`",
        "text": "[Links]",
        "type": "code",
      },
      {
        "raw": " or ",
        "text": " or ",
        "type": "plain",
      },
      {
        "raw": "\`#links\`",
        "text": "#links",
        "type": "code",
      },
      {
        "raw": " are two ways to make links. They do three things",
        "text": " are two ways to make links. They do three things",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 2,
    "nodes": [
      {
        "raw": "An internal link to a page",
        "text": "An internal link to a page",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 2,
    "nodes": [
      {
        "raw": "A ",
        "text": "A ",
        "type": "plain",
      },
      {
        "content": "",
        "href": "Bi-directional link",
        "pathType": "relative",
        "raw": "[Bi-directional link]",
        "type": "link",
      },
      {
        "raw": " back to the source",
        "text": " back to the source",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 2,
    "nodes": [
      {
        "raw": "A ",
        "text": "A ",
        "type": "plain",
      },
      {
        "content": "",
        "href": "2-hop link",
        "pathType": "relative",
        "raw": "[2-hop link]",
        "type": "link",
      },
      {
        "raw": " so you can find more related pages",
        "text": " so you can find more related pages",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": "Block quote",
            "text": "Block quote",
            "type": "plain",
          },
        ],
        "raw": "[[Block quote]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "raw": " use the right caret ",
            "text": " use the right caret ",
            "type": "plain",
          },
          {
            "raw": "\`>\`",
            "text": ">",
            "type": "code",
          },
          {
            "raw": " at the beginning of a line to get a block quote ",
            "text": " at the beginning of a line to get a block quote ",
            "type": "plain",
          },
        ],
        "raw": "> use the right caret \`>\` at the beginning of a line to get a block quote ",
        "type": "quote",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "raw": "Mouse based styling",
            "text": "Mouse based styling",
            "type": "plain",
          },
        ],
        "raw": "[* Mouse based styling]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "link": "",
        "raw": "[https://gyazo.com/a515ab169b1e371641f7e04bfa92adbc]",
        "src": "https://gyazo.com/a515ab169b1e371641f7e04bfa92adbc/thumb/1000",
        "type": "image",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "content": "",
            "href": "Code notation",
            "pathType": "relative",
            "raw": "[Code notation]",
            "type": "link",
          },
        ],
        "raw": "[[[Code notation]]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Use backquotes or backticks, \`,  to highlight code  ",
        "text": "Use backquotes or backticks, \`,  to highlight code  ",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "e.g. ",
        "text": "e.g. ",
        "type": "plain",
      },
      {
        "raw": "\`function() {  return true }\`",
        "text": "function() {  return true }",
        "type": "code",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "content": "",
            "href": "Code blocks",
            "pathType": "relative",
            "raw": "[Code blocks]",
            "type": "link",
          },
        ],
        "raw": "[[[Code blocks]]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Typing ",
        "text": "Typing ",
        "type": "plain",
      },
      {
        "raw": "\`code:filename.extension\`",
        "text": "code:filename.extension",
        "type": "code",
      },
      {
        "raw": "or",
        "text": "or",
        "type": "plain",
      },
      {
        "raw": "\`code:filename\`",
        "text": "code:filename",
        "type": "code",
      },
      {
        "raw": "can be used to create a new code snippet and and display it as a block",
        "text": "can be used to create a new code snippet and and display it as a block",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 2,
    "nodes": [
      {
        "raw": "Language names may be abbreviated",
        "text": "Language names may be abbreviated",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "content": "function () {
  alert(document.location.href)
  console.log("hello")
  // You can also write comments!
}",
    "fileName": "hello.js",
    "indent": 1,
    "type": "codeBlock",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "nodes": [
          {
            "content": "",
            "href": "Tables",
            "pathType": "relative",
            "raw": "[Tables]",
            "type": "link",
          },
        ],
        "raw": "[[[Tables]]]",
        "type": "strong",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Type table: tablename to create a table",
        "text": "Type table: tablename to create a table",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Use tab to move to the next column, use enter to move to the next row.",
        "text": "Use tab to move to the next column, use enter to move to the next row.",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "An example:",
        "text": "An example:",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "cells": [
      [
        [
          {
            "raw": "1",
            "text": "1",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "2",
            "text": "2",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      [
        [
          {
            "raw": "1 ",
            "text": "1 ",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "2 ",
            "text": "2 ",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "3",
            "text": "3",
            "type": "plain",
          },
        ],
      ],
      [
        [
          {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "------",
            "text": "------",
            "type": "plain",
          },
        ],
      ],
      [
        [
          {
            "raw": "a",
            "text": "a",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "b",
            "text": "b",
            "type": "plain",
          },
        ],
        [
          {
            "raw": "c",
            "text": "c",
            "type": "plain",
          },
        ],
      ],
    ],
    "fileName": "hello",
    "indent": 0,
    "type": "table",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "content": "",
            "href": "Mathematical notation",
            "pathType": "relative",
            "raw": "[Mathematical notation]",
            "type": "link",
          },
        ],
        "raw": "[* [Mathematical notation]]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "Using ",
        "text": "Using ",
        "type": "plain",
      },
      {
        "content": "TeX",
        "href": "https://en.wikipedia.org/wiki/TeX",
        "pathType": "absolute",
        "raw": "[TeX https://en.wikipedia.org/wiki/TeX]",
        "type": "link",
      },
      {
        "raw": " inside of brackets with a dollar sign ",
        "text": " inside of brackets with a dollar sign ",
        "type": "plain",
      },
      {
        "raw": "\`[$ TeX here ]\`",
        "text": "[$ TeX here ]",
        "type": "code",
      },
      {
        "raw": ", you can format math or science formulas, like so: ",
        "text": ", you can format math or science formulas, like so: ",
        "type": "plain",
      },
      {
        "formula": "E = mc^2",
        "raw": "[$ E = mc^2]",
        "type": "formula",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [
      {
        "decos": [
          "*-1",
        ],
        "nodes": [
          {
            "content": "",
            "href": "Userscript",
            "pathType": "relative",
            "raw": "[Userscript]",
            "type": "link",
          },
        ],
        "raw": "[* [Userscript]]",
        "rawDecos": "*",
        "type": "decoration",
      },
    ],
    "type": "line",
  },
  {
    "indent": 1,
    "nodes": [
      {
        "raw": "You can even add javascript to customize Scrapbox to your liking.",
        "text": "You can even add javascript to customize Scrapbox to your liking.",
        "type": "plain",
      },
    ],
    "type": "line",
  },
  {
    "indent": 0,
    "nodes": [],
    "type": "line",
  },
]
`;
